### ISSUE TITLE
Custom attributes, additional interfaces, and constructor parameters unused when we create a Fake delegate

### ISSUE THREAD

**Issue by blairconrad** [FakeItEasy/FakeItEasy#1353]
Mentioned by @thomaslevesque in https://github.com/FakeItEasy/FakeItEasy/pull/1352#discussion_r187769671

We should

1. document this, as [Creating Fakes](http://fakeiteasy.readthedocs.io/en/stable/creating-fakes/#explicit-creation-options) does not mention it, and
2. make sure we throw a nice exception if attributes are supplied during creation

**Comment by thomaslevesque** [comment]
We should also investigate if it's possible at all to add attributes to fake delegates. Would that even make sense? I'm not sure, my brain is a bit slow today.

**Comment by thomaslevesque** [comment]
> Would that even make sense?

It wouldn't make sense to apply attributes to the delegate type (because we're not creating a new type, it already exists) or to the delegate instance (because instances are not reflection objects, they don't have attributes). It might make sense to apply attributes to the method that the delegate instance encapsulates. I can't think of many real-world use cases, though.

> We should also investigate if it's possible at all to add attributes to fake delegates.

I don't think it's possible, at least with our current approach to generate fake delegates. Using [`CompileToMethod`](https://docs.microsoft.com/en-us/dotnet/api/system.linq.expressions.lambdaexpression.compiletomethod?view=netframework-4.7.2#System_Linq_Expressions_LambdaExpression_CompileToMethod_System_Reflection_Emit_MethodBuilder_) instead of `Compile` might make it possible to apply attributes to the method, but `CompileToMethod` isn't available in .NET Core.

**Comment by blairconrad** [comment]
@thomaslevesque, after your last comment, I think we're left with "just document the behaviour and throw when attributes are supplied on delegate creation", no?

**Comment by thomaslevesque** [comment]
> @thomaslevesque, after your last comment, I think we're left with "just document the behaviour and throw when attributes are supplied on delegate creation", no?

Yes, I think so.

**Comment by blairconrad** [comment]
Throwing when the user tries to add attributes to a delegate is easy, if we want to do it while trying constructors (of which I'd expect only one). We can elicit this error:

```
  Failed to create fake of type System.Func`1[System.Int32]:
   Below is a list of reasons for failure per attempted constructor:
    Constructor with signature (System.Object, System.IntPtr) failed:
      Cannot add custom attributes to faked delegates.
```

If we like that, I can send a PR real soon now.
I did start chasing a pre-constructor based rejection, but it starts to spread, not entirely unlike the recent "why can't we Dummy?" issue.
This is mostly due to the current practice of saying "if a delegate proxy generator isn't appropriate, then we can use the dynamic proxy generator".

Still, if there's perceived value in rejecting the construction without trying constructors, I'll carry on!

**Comment by thomaslevesque** [comment]
> if we want to do it while trying constructors

I'm not sure which constructors you're talking about... We don't "try constructors" when faking delegates.

Couldn't we check for this in `IFakeOptions.WithAttributes`?

**Comment by thomaslevesque** [comment]
> We could, if we wanted to spread knowledge of the various proxy types and what they can handle around the system. Right now it's a little more concentrated in the creation code.

Good point. So the proper place to check this would probably be `DelegateProxyGenerator.GenerateProxy`, no?

BTW, I just realized that it's not just the attribute list: we also ignore `additionalInterfacesToImplement` and `argumentsForConstructor`.

**Comment by blairconrad** [comment]
> BTW, I just realized that it's not just the attribute list: we also ignore additionalInterfacesToImplement and argumentsForConstructor.

Yup. I noticed `additionalInterfacesToImplement` last night and `argumentsForConstructor` this morning as a result of your comment.

> Good point. So the proper place to check this would probably be DelegateProxyGenerator.GenerateProxy, no?

I think so. So proper. Somewhat annoying. Okay. I'll keep plugging!

**Comment by blairconrad** [comment]
> Good point. So the proper place to check this would probably be DelegateProxyGenerator.GenerateProxy, no?

Oh. That actually might work out easier than I was thinking. I'd thunk of `DelegateProxyGenerator.CanGenerateProxy`. But delaying until `GenerateProxy` might significantly simplify things! Thanks!

**Comment by blairconrad** [comment]
Ha. I'm foolish. I'd originally put the check in `GenerateProxy`, which is what gives us the error I pasted way upstairs. The one that yammers on about "constructors".

**Comment by thomaslevesque** [comment]
Ah. So maybe we need to special case delegates to not mention constructors...

**Comment by blairconrad** [comment]
I'm sensing an "ResolveStrategy" for Fakes to match the one we have for Dummies. :)

That might actually not be a bad approach. `MethodCanBeInterceptedOnInstance` might cause a hiccough, though. Or maybe not…

**Comment by thomaslevesque** [comment]
Maybe not a bad approach, but probably overkill...
For now, maybe it's easier to just check in `IFakeOptions.WithAttributes`, although it's not ideal. No sense in upsetting everything to fix an issue that probably isn't bothering anyone.

**Comment by blairconrad** [comment]
Oh, but I'm having fun. Let me go a bit longer and then I can send a PR and you can reject it and I'll do the `IFakeOptions.WithAttributes` thing!

**Comment by thomaslevesque** [comment]
Lol! OK, have a go at it then ;)

**Comment by blairconrad** [comment]
Relabelling as a bug, since we're no longer planning on supporting the attributes (interfaces, constructor arguments…)

**Comment by blairconrad** [comment]
This change has been released as part of [FakeItEasy 4.8.0](https://github.com/FakeItEasy/FakeItEasy/releases/4.8.0).
